### PR TITLE
First pass at pytorch.nn.Module serializer

### PR DESCRIPTION
**Pull request by dmaljovec**
~This PR implements the [documented best practice](https://pytorch.org/docs/stable/notes/serialization.html#best-practices) for serializing a pytorch module.~

~So, I am not fully happy with this just yet. For one, I do need to add tests for this. Secondly, in order to load a model you need to know what type of model was saved. I could wrap this in the saved object somehow and reconstruct it back out, or I could use the `not best practice` method of saving the entire model rather than just the `state_dict`? I am not entirely sure how this is being used in practice yet, and maybe this is a non-issue because it is easy enough to pass the `model_class` into the deserializer?~

~Anyway, opening a draft PR now to give some visibility as to what I am doing. The last question I have is whether we want to serialize the base torch Module, or something in the lightning package? It would seem to me that somewhere in lightning it is using a base pytorch module that we could write a thin wrapper to that will call the serializer I have here underneath the covers. The open question here is whether there is anything worth saving in the lightning layer of convenience?~

This PR does not implement the [documented best practice](https://pytorch.org/docs/stable/notes/serialization.html#best-practices) for serializing a pytorch module since doing so implies you can infer what the model type is on load. Instead we write the entire model to file, and load the entire model on deserialization. This will break on changes of directory structure and has potential to give unexpected results on updates to a model given how PyTorch serializes models. It stores the file path of a model when it is saved, so any changes to an existing model will be dynamically loaded on deserialization which means updates to a model will be available on load, this could be a pleasant feature or a particularly nasty bug to unwind if a user expects the result of the original model.

### CODE REVIEW

File: tests/provenance/test_pytorch.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+
+def test_same_seeds_result_in_same_models(dbdiskrepo):
+    data = random_data()
+
+    torch.manual_seed(0)
+    model1 = basic_model()
+    fit1 = fit_model(model1, data['X_train'], data['Y_train'])
+
+    torch.manual_seed(0)
+    model2 = basic_model()
```
Comment (bmabey) [inline tests/provenance/test_pytorch.py]: @dmaljovec At this point does `hash(model1) == hash(model2)`

Comment (bmabey) [reply tests/provenance/test_pytorch.py]: Oh wait, I see they that below you assert that have the same artifact ids so they should have the same hash.

**Comment by dmaljovec** [comment]
Okay, @bmabey I think this may be ready for review? I want to get this nailed down and then will do a `pytorch-lightning` serializer as a follow-on PR.

**Comment by dmaljovec** [comment]
Talked offline, it looks like supporting a `pytorch-lightning` serializer is not likely to happen in the near-term, due to how that project is structured around logging/checkpointing.

`Trainer.fit` does not return a model, so we would have to be looking into the last checkpoint and grabbing the torch model for serialization, a burden that would have to fall on the developer using provenance.

**Comment by jesford** [comment]
@bmabey could we go ahead and merge this? I've been using @dmaljovec 's fork and it has unblocked me on using `provenance` with new `wrapt` versions.

**Comment by bmabey** [comment]
@jesford Yes, sorry, I thought this had been merged. It looks like there are conflicts that need to be resolved. I can look into it tomorrow or if you or @dmaljovec resolve them I can merge them in sooner.

**Comment by dmaljovec** [comment]
I do not see conflicts on my end. Which files are in conflict for you?

**Comment by bmabey** [comment]
image
@dmaljovec ^

**Comment by dmaljovec** [comment]
Screen Shot 2019-11-27 at 6 53 01 AM

This is strange indeed. Let me pull your master locally and see if there is anything I can resolve locally.

**Comment by dmaljovec** [comment]
So, I tried to ensure I am on top of the latest master branch via the following commands:

```
$ git remote -v
origin  git@github.com:dmaljovec/provenance.git (fetch)
origin  git@github.com:dmaljovec/provenance.git (push)
upstream        git@github.com:bmabey/provenance.git (fetch)
upstream        git@github.com:bmabey/provenance.git (push)
$ git pull upstream master
From github.com:bmabey/provenance
 * branch            master     -> FETCH_HEAD
Already up to date.
$ git checkout pytorch-serializer 
Already on 'pytorch-serializer'
Your branch is up to date with 'origin/pytorch-serializer'.
$ git merge master
Already up to date.
```

I could open a new branch that builds off the latest master and applies these same changes, but not sure how to get this branch to signal that it is not in conflict?

**Comment by dmaljovec** [comment]
I opened #57 which is the same exact changes, but in a single commit that was constructing using the latest master branch and should not have any conflicts.

**Pull request closed without merging**